### PR TITLE
[browser][MT] pre-load threads early in dotnet.js

### PR DESCRIPTION
**Pull request by pavelsavara**
- pre-load threads early in `dotnet.js`
    - they are merged into emscripten pool in new `populateEmscriptenPool`
- introduce new `workerNumber` sequence
    - send `monoThreadInfo` from UI to worker together with `config`
- introduce `pthreadPoolReady` config as floating pool and keep `pthreadPoolSize` as pre-init pool. See `getNewWorker`
- moved `mono_wasm_init_diagnostics` bit later, when thread pool is ready, but before mono start
- moved `init_finalizer_thread` even later, 200ms after runtime is ready. This may improve [GC deadlock ](https://github.com/dotnet/runtime/issues/96190)
- refactor how we handle emscripten internals, separated UI and workers
- move thread typescript types to `types\internal.ts`

### CODE REVIEW

File: src/mono/browser/runtime/diagnostics/index.ts [hunk 1]
```diff
@@ -37,12 +37,9 @@ let diagnosticsServerEnabled = false;
 let diagnosticsInitialized = false;
 
 export async function mono_wasm_init_diagnostics(): Promise<void> {
-    if (diagnosticsInitialized)
-        return;
-    if (!WasmEnableThreads) {
-        mono_log_warn("ignoring diagnostics options because this runtime does not support diagnostics");
```
Comment (maraf) [inline src/mono/browser/runtime/diagnostics/index.ts:43]: nit: I think any similar clue is valuable for the customers

File: src/mono/browser/runtime/exports.ts [hunk 2]
```diff
@@ -70,6 +70,10 @@ function initializeExports(globalObjects: GlobalObjects): RuntimeAPI {
         runtimeList = globalThisAny.getDotnetRuntime.__list;
     }
 
+    if (BuildConfiguration === "Debug") {
+        globalThisAny.INTERNAL = globals.internal;
```
Comment (maraf) [inline src/mono/browser/runtime/exports.ts:74]: Could you please add an explanation?
Would this break our test with multiple runtimes?
Could it based on some parameter?

File: src/mono/browser/runtime/loader/assets.ts [hunk 3]
```diff
@@ -733,4 +733,24 @@ export async function streamingCompileWasm() {
...
+        const workerNumber = loaderHelpers.workerNextNumber++;
+        const worker: Partial<PThreadWorker> = new Worker(jsModuleWorker.resolvedUrl!, {
+            name: "dotnet-worker-" + workerNumber.toString().padStart(3, "0"),
+        });
+        worker.info = {
+            workerNumber,
+            pthreadId: PThreadPtrNull,
+            reuseCount: 0,
+            updateCount: 0,
+            threadPrefix: "          -    ",
```
Comment (maraf) [inline src/mono/browser/runtime/loader/assets.ts:751]: Can we have the naming unified?

File: src/mono/browser/runtime/loader/config.ts [hunk 4]
```diff
@@ -187,10 +187,13 @@ export function normalizeConfig() {
         config.cachedResourcesPurgeDelay = 10000;
     }
 
+    // ActiveIssue https://github.com/dotnet/runtime/issues/75602
```
Comment (maraf) [inline src/mono/browser/runtime/loader/config.ts:190]: Unrelated, but shouldn't this come from the tests configuration?

File: src/mono/browser/runtime/pthreads/browser/index.ts [hunk 5]
```diff
@@ -140,35 +135,26 @@ export function thread_available(): Promise<void> {
     return pendingWorkerLoad.promise;
 }
 
+export function populateEmscriptenPool(): void {
+    if (!WasmEnableThreads) return;
+    const unused = getUnusedWorkerPool();
+    for (const worker of loaderHelpers.loadingWorkers) {
+        unused.push(worker);
+    }
+    loaderHelpers.loadingWorkers.length = 0; // GC
```
Comment (maraf) [inline src/mono/browser/runtime/pthreads/browser/index.ts:144]: Never seen this, is it enough in all browsers?

File: src/mono/browser/runtime/pthreads/browser/index.ts [hunk 6]
```diff
@@ -217,3 +206,16 @@ export function dumpThreads(): void {
...
+    // we don't need it immediately, so we can wait a bit, to keep CPU working on normal startup
+    setTimeout(() => {
+        try {
+            cwraps.mono_wasm_init_finalizer_thread();
+        }
+        catch (err) {
+            mono_log_error("init_finalizer_thread() failed", err);
+            loaderHelpers.mono_exit(1, err);
+        }
+    }, 200);
```
Comment (maraf) [inline src/mono/browser/runtime/pthreads/browser/index.ts:220]: Can we do better that fixed number? I guess mobile phone VS our dev machines make difference

File: src/mono/browser/runtime/types/index.ts [hunk 7]
```diff
@@ -144,6 +144,10 @@ export type MonoConfig = {
      * initial number of workers to add to the emscripten pthread pool
      */
     pthreadPoolSize?: number,
+    /**
+     * initial number of unused workers keep in the emscripten pthread pool after startup
+     */
+    pthreadPoolReady?: number,
```
Comment (maraf) [inline src/mono/browser/runtime/types/index.ts:150]: I would prefer a different name

- pthreadPoolInitSize
- workerPoolInitSize
- ...

File: src/mono/browser/runtime/types/index.ts [hunk 8]
```diff
@@ -144,6 +144,10 @@ export type MonoConfig = {
      * initial number of workers to add to the emscripten pthread pool
      */
     pthreadPoolSize?: number,
+    /**
+     * initial number of unused workers keep in the emscripten pthread pool after startup
```
Comment (maraf) [inline src/mono/browser/runtime/types/index.ts:148]:
```suggestion
     * initial number of unused workers kept in the emscripten pthread pool after startup
```

File: src/mono/browser/runtime/loader/config.ts (hunk 4 above)
Comment (pavelsavara) [reply src/mono/browser/runtime/loader/config.ts:190]: You can override it in tests, but at the moment we don't. The issue is to figure out what is the default for customers and then we bump the number for offending tests.

File: src/mono/browser/runtime/types/index.ts (hunk 7 above)
Comment (pavelsavara) [reply src/mono/browser/runtime/types/index.ts:150]: `pthreadPoolInitialSize`, `pthreadPoolUnusedSize`

File: src/mono/browser/runtime/pthreads/browser/index.ts (hunk 6 above)
Comment (pavelsavara) [reply src/mono/browser/runtime/pthreads/browser/index.ts:220]: `finalizerThreadStartDelayMs`

File: src/mono/browser/runtime/exports.ts (hunk 2 above)
Comment (pavelsavara) [reply src/mono/browser/runtime/exports.ts:74]: BuildConfiguration === "Debug"

Comment (pavelsavara) [reply src/mono/browser/runtime/exports.ts:74]: anyway, I will remove it

File: src/mono/browser/runtime/pthreads/shared.ts [hunk 9]
```diff
@@ -4,15 +4,24 @@
...
+import type { GCHandle, MonoThreadMessage, PThreadInfo, PThreadPtr } from "../types/internal";
 
-/// Messages sent on the dedicated mono channel between a pthread and the browser thread
+import { ENVIRONMENT_IS_PTHREAD, Module, loaderHelpers, mono_assert, runtimeHelpers } from "../globals";
+import { set_thread_prefix } from "../logging";
+import { bindings_init } from "../startup";
+import { forceDisposeProxies } from "../gc-handles";
+import { monoMessageSymbol, GCHandleNull, PThreadPtrNull, WorkerToMainMessageType } from "../types/internal";
+
+export const worker_empty_prefix = "          -    ";
```
Comment (maraf) [inline src/mono/browser/runtime/pthreads/shared.ts:15]:
```suggestion
// A duplicate in loader/assets.ts
export const worker_empty_prefix = "          -    ";
```

File: src/mono/browser/runtime/loader/assets.ts [hunk 10]
```diff
@@ -20,6 +20,8 @@ let parallel_count = 0;
 const assetsToLoad: AssetEntryInternal[] = [];
 const singleAssets: Map<string, AssetEntryInternal> = new Map();
 
+const worker_empty_prefix = "          -    ";
```
Comment (maraf) [inline src/mono/browser/runtime/loader/assets.ts:23]:
```suggestion
// A duplicate in pthreads/shared.ts
const worker_empty_prefix = "          -    ";
```

File: src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/BootJsonData.cs [hunk 11]
```diff
@@ -108,9 +108,14 @@ public class BootJsonData
     public object diagnosticTracing { get; set; }
 
     /// <summary>
-    /// Gets or sets pthread pool size.
+    /// Gets or sets pthread pool initial size.
     /// </summary>
-    public int? pthreadPoolSize { get; set; }
+    public int? pthreadPoolInitialSize { get; set; }
```
Comment (maraf) [inline src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/BootJsonData.cs:113]: Please open an issue we need to implement these in WebAssembly SDK and assign it to me

**Review by maraf — APPROVED** [review]

**Pull request merged**